### PR TITLE
ROX-25137: Added nvd cvss field for policy

### PR DESCRIPTION
**Pull request by ksurabhi91**
This PR adds nvd cvss field for policy criteria. 
Tested policy violations for nvd cvss and results are in this doc https://docs.google.com/document/d/1EvLIZjdmJF55KifOIYzevEZF_pdjFJ_JUgmb5MiPx2o/edit
Tested the change by calling  the PUT API to add nvd cvss filed to a user defined policy and verifying violations generated by the policy
Screenshot 2024-10-11 at 12 16 55 PM
Screenshot 2024-10-11 at 12 17 15 PM
Screenshot 2024-10-16 at 3 57 53 PM

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

change me!

### CODE REVIEW

File: proto/storage/vulnerability.proto [hunk 1]
```diff
@@ -55,6 +55,7 @@ message EmbeddedVulnerability {
   VulnerabilityState state = 20; // @gotags: search:"Vulnerability State"
   // cvss_metrics stores list of cvss scores from different sources like nvd, Redhat etc
   repeated CVSSScore cvss_metrics = 21;
+  float nvd_cvss = 22; // @gotags: search:"NVD CVSS,store"
```
Comment (clickboo) [inline proto/storage/vulnerability.proto:58]: will need the policy tag here

File: central/image/datastore/store/postgres/store.go [hunk 2]
```diff
@@ -360,8 +361,12 @@ func copyFromImageCves(ctx context.Context, tx *postgres.Tx, iTime time.Time, ob
 	if err != nil {
 		return err
 	}
+	// for testing only
+	r := rand.New(rand.NewSource(99))
```
Comment (clickboo) [inline central/image/datastore/store/postgres/store.go:365]: Assuming based on the comment, you plan to revert this before merging.

File: pkg/booleanpolicy/violationmessages/printer.go [hunk 3]
```diff
@@ -59,6 +59,7 @@ var (
 		fieldnames.HasEgressNetworkPolicy:         {{required: set.NewStringSet(augmentedobjs.HasEgressPolicyCustomTag), printerFuncKey: printer.HasEgressNetworkPolicyKey}},
 		fieldnames.MountPropagation:               {{required: set.NewStringSet(search.MountPropagation.String()), printerFuncKey: printer.VolumeKey}},
 		fieldnames.Namespace:                      {{required: set.NewStringSet(search.Namespace.String()), printerFuncKey: printer.NamespaceKey}},
+		fieldnames.NvdCvss:                        {{required: set.NewStringSet(search.CVE.String()), printerFuncKey: printer.NvdcveKey}},
```
Comment (clickboo) [inline pkg/booleanpolicy/violationmessages/printer.go:62]: nit: naming: NvdCveKey (C capital)

File: pkg/booleanpolicy/violationmessages/printer/vuln.go [hunk 4]
```diff
@@ -24,6 +24,12 @@ const (
     {{- if .FixedBy}}Fixable {{end}}{{.CVE}}{{if .CVSS}} (CVSS {{.CVSS}}){{end}}{{if .Severity}} (severity {{.Severity}}){{end}} found
```
Comment (clickboo) [inline pkg/booleanpolicy/violationmessages/printer/vuln.go:24]: I don't think this should be an either or template - rather two separate templates one for cvss and one for nvd. The nvd score can be added, comma separated from the cvss score in the parens, in the same message.

File: pkg/booleanpolicy/violationmessages/printer/vuln.go [hunk 5]
```diff
@@ -24,6 +24,12 @@ const (
...
     {{- if .Component}} in component '{{.Component}}' (version {{.ComponentVersion}}){{end}}
     {{- if .ContainerName }} in container '{{.ContainerName}}'{{end}}
+    {{- if .FixedBy}}, resolved by version {{.FixedBy}}{{end}}`
+
+	nvdcveTemplate = `
```
Comment (clickboo) [inline pkg/booleanpolicy/violationmessages/printer/vuln.go:29]: If indeed two templates are needed, please add a sample message in the comments like the template above.

Comment (ksurabhi91) [reply pkg/booleanpolicy/violationmessages/printer/vuln.go:29]: @clickboo  Here is sample message from unit test
CVE-2014-0160 (NVD CVSS 8) (severity Unknown) found in component 'heartbleed' (version 1.2) in container 'nginx'

**Comment by clickboo** [comment]
Would be nice to get screenshots of the violation messages in the violations -> events tab

File: proto/storage/cve.proto [hunk 6]
```diff
@@ -148,7 +148,7 @@ message ImageCVE {
   google.protobuf.Timestamp snooze_start = 8 [deprecated = true];
   google.protobuf.Timestamp snooze_expiry = 9 [deprecated = true]; // @gotags: search:"CVE Snooze Expiry,hidden"
   // nvdcvss stores cvss score for a cve from NVD
-  float nvdcvss = 10; // @gotags: search:"NVD CVSS",store"
+  float nvdcvss = 10; // @gotags: search:"NVD CVSS,store"
```
Comment (charmik-redhat) [inline proto/storage/cve.proto:151]: Does that quote change anything? The search field worked before right?

Comment (ksurabhi91) [reply proto/storage/cve.proto:151]: yes it worked before with graphql query.

File: pkg/booleanpolicy/default_policies_test.go [hunk 7]
```diff
@@ -110,6 +110,55 @@ func (suite *DefaultPoliciesTestSuite) imageIDFromDep(deployment *storage.Deploy
...
+		},
+	}
+
+	policy := policyWithGroups(storage.EventSource_NOT_APPLICABLE, nvdCvssPolicyGroup)
+
+	deployment := suite.deployments["HEARTBLEEDDEPID"]
+	depMatcher, err := BuildDeploymentMatcher(policy)
+	require.NoError(suite.T(), err)
+	violations, err := depMatcher.MatchDeployment(nil, enhancedDeployment(deployment, suite.getImagesForDeployment(deployment)))
+	require.Len(suite.T(), violations.AlertViolations, 1)
```
Comment (charmik-redhat) [inline pkg/booleanpolicy/default_policies_test.go:157]: Can you assert the violation message here too? Or is it not available here?

File: proto/storage/vulnerability.proto (hunk 1 above)
Comment (charmik-redhat) [inline proto/storage/vulnerability.proto:58]: Does it not need a policy tag too? Or is search tag enough?

Comment (ksurabhi91) [reply proto/storage/vulnerability.proto:58]: @charmik-redhat  cvss only has a search tag. based on testing so far search tag looks enough

File: central/image/datastore/store/postgres/store_test.go [hunk 8]
```diff
@@ -44,6 +44,11 @@ func (s *ImagesStoreSuite) TestStore() {
 
 	image := fixtures.GetImage()
 	s.NoError(testutils.FullInit(image, testutils.SimpleInitializer(), testutils.JSONFieldsFilter))
+	for _, comp := range image.GetScan().GetComponents() {
+		for _, vuln := range comp.GetVulns() {
+			vuln.NvdCvss = 0
```
[the remark this answers is not in the document]
Comment (ksurabhi91) [reply central/image/datastore/store/postgres/store_test.go:49]: test assigns rnadom value to nvd cvss. just resetting it to 0 here

**Review by clickboo — APPROVED** [review]
lgtm, 🚢 after CI passes. Assuming the new violation printer message was via manual test.

**Review by charmik-redhat — APPROVED** [review]

**Pull request merged**